text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+  * [Traffic Balancing Optimization](#traffic-balancing-optimization)
+  * [Throttling](#throttling)
+  * [Best Effort Delivery Guarantee](#best-effort-delivery-guarantee)
+* [Risks and Mitigation](#risks-and-mitigations)
+* [Trade-offs and Mitigations](#trade-offs-and-mitigations)
+  * [Duplicate Data](#duplicate-data)
+  * [Incompatible Backends](#incompatible-backends)
+  * [Small Devices/Small Telemetry Data Stream](#small-devicessmall-telemetry-data-stream)
+* [Future Versions and Interoperability](#future-versions-and-interoperability)
+* [Prior Art and Alternatives](#prior-art-and-alternatives)
+* [Open Questions](#open-questions)
+* [Appendix A - Protocol Buffer Definitions](#appendix-a---protocol-buffer-definitions)
+* [Appendix B - Performance Benchmarks](#appendix-b---performance-benchmarks)
+* [Appendix C - Parameter Tuning and Design Optimization](#appendix-c---parameter-tuning-and-design-optimization)
+* [Glossary](#glossary)
+* [Acknowledgements](#acknowledgements)
+
+## Introduction
+
+### Motivation
+
+As telemetry data becomes more widely available and volumes increase, new uses and needs are emerging for the OTLP
+ecosystem: cost-effectiveness, advanced data processing, data minimization. This OTEP aims to extend the OTLP
+protocol to better address them while maintaining the compatibility with the existing ecosystem.
+
+Currently, the OTLP protocol uses a "row-oriented" format to represent all the OTEL entities. This representation works
+well for small batches (<50 entries) but, as the analytical database industry has shown, a "column-oriented"
+representation is more optimal for the transfer and processing of *large batches* of entities. The term "row-oriented"
+is used when data is organized into a series of records, keeping all data associated with a record next to each other in
+memory. A "column-oriented" system organizes data by fields, grouping all the data associated with a field next to each
+other in memory. The main benefits of a columnar approach are:
+
+* **better data compression rate** (arrays of similar data generally compress better),
+* **faster data processing** (see diagram below),
+* **faster serialization and deserialization** (few arrays vs many in-memory objects to serialize/deserialize),
+* **better IO efficiency** (less data to transmit).
+
+![row vs column-oriented](img/0156_OTEL%20-%20Row%20vs%20Column.png)
+
+This OTEP proposes to extend the [OpenTelemetry protocol (OTEP 0035)](https://github.com/open-telemetry/oteps/blob/main/text/0035-opentelemetry-protocol.md)
+with a **generic columnar representation for metrics, logs and traces based on Apache Arrow**. Compared to the existing
+OpenTelemetry protocol this compatible extension has the following improvements:
+
+* **Reduce the bandwidth requirements** of the protocol. The two main levers are: 1) a better representation of the
+  telemetry data based on a columnar representation, 2) a stream-oriented gRPC endpoint that is more efficient to
```

Comment:
Was there a benchmark that illustrates that stream-oriented gRPC endpoint is beneficial vs. a simpler req/resp? Tigran did a similar analysis for the original OTLP and concluded that streaming did not provide significant perf benefits, but had obvious operational downsides (load balancing, back pressure, retries).

Reply:
The point of this bullet is that we intend to measure this and we are aware of reasons why some users will prefer it even at a loss of compression. We don't see any reason that unary RPCs won't work, and we don't think that will perform worse than standard OTLP. This shouldn't be a blocker.

We are tracking this [here](https://github.com/open-telemetry/experimental-arrow-collector/issues/36), it's a frequent question and one I'll be glad to answer. I think users will demand a unary RPC form of this no matter what, especially so that we can use plain HTTP instead of gRPC. I expect that there is batch size above which the benefit of streaming RPC is negligible and if users are willing to (a) handle such large batches and (b) have enough traffic to avoid significant delays, then maybe they'll switch to unary RPCs to simplify load balancing.